### PR TITLE
WIP: Actor for performing rollups for provided metrics against KairosDB.

### PR DESCRIPTION
**Pull request by ddimensia**

### CODE REVIEW

**Comment by ddimensia** [comment]
I just wanted to get some initial eyes on this.  I still need to wire the actor into Guice and deal with metrics but this is mostly complete.

File: app/com/arpnetworking/rollups/RollupGenerator.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+                .build());
+    }
+
+    private final ActorRef _metricsDiscovery;
+    private final KairosDbClient _kairosDbClient;
+    private final int _maxBackFillPeriods;
+    private final FiniteDuration _fetchBackoff;
+    private final Clock _clock;
+    private List<RollupPeriod> _periodsInFlight = Collections.emptyList();
+    static final Object FETCH_METRIC = new Object();
```
Comment (vjkoskela) [inline app/com/arpnetworking/rollups/RollupGenerator.java]: nit: convention is to have a blank line between non-static and static members.

File: app/com/arpnetworking/kairos/client/models/Aggregator.java [hunk 2]
```diff
@@ -41,30 +41,46 @@ private Aggregator(final Builder builder) {
             _alignSampling = builder._alignSampling;
         }
         _otherArgs = builder._otherArgs;
+        _alignEndTime = builder._alignEndTime;
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/kairos/client/models/Aggregator.java]: fixed

File: app/com/arpnetworking/kairos/client/models/Aggregator.java [hunk 3]
```diff
@@ -115,6 +131,30 @@ public Builder setAlignSampling(final Optional<Boolean> value) {
...
         }
 
+        /**
+         * Sets the align_start_time of the aggregator. Optional. Defaults to false.
+         *
+         * @param value the align_start_time for the aggregator
+         * @return this {@link Builder}
+         */
+        @JsonProperty("align_start_time")
+        public Builder setAlignStartTime(final Optional<Boolean> value) {
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/kairos/client/models/Aggregator.java]: fixed

File: app/com/arpnetworking/kairos/client/models/Aggregator.java [hunk 4]
```diff
@@ -145,7 +185,11 @@ public Builder setOtherArgs(final ImmutableMap<String, Object> value) {
         @NotNull
         private Optional<Boolean> _alignSampling = Optional.of(true);
         @NotNull
-        private Optional<Sampling> _sampling = Optional.of(new Sampling.Builder().build());
+        private Optional<Sampling> _sampling = Optional.empty();
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/kairos/client/models/Aggregator.java]: fixed

File: app/com/arpnetworking/kairos/client/models/Metric.java [hunk 5]
```diff
@@ -50,17 +53,31 @@ public String getName() {
         return _groupBy;
     }
 
+    @JsonInclude(JsonInclude.Include.NON_NULL)
+    public Integer getLimit() {
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/kairos/client/models/Metric.java]: fixed

File: app/com/arpnetworking/rollups/LastDataPointMessage.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+import com.google.common.collect.ImmutableSet;
+import net.sf.oval.constraint.NotEmpty;
+import net.sf.oval.constraint.NotNull;
+
+import java.time.Instant;
+import java.util.Optional;
+import javax.annotation.Nullable;
+
+/**
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/rollups/LastDataPointMessage.java]: fixed

File: test/com/arpnetworking/rollups/RollupGeneratorTest.java [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+import static org.mockito.internal.verification.VerificationModeFactory.times;
+
+/**
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply test/com/arpnetworking/rollups/RollupGeneratorTest.java]: fixed

File: app/com/arpnetworking/rollups/RollupGenerator.java [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+        PatternsCS.pipe(_kairosDbClient.queryMetricTags(
+                new MetricsQuery.Builder()
+                        .setStartTime(Instant.ofEpochMilli(0))
+                        .setMetrics(ImmutableList.of(
+                                new Metric.Builder()
+                                        .setName(metricName)
+                                        .build()
+                        ))
+                        .build()).handle((response, failure) -> {
+                    if (failure != null) {
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/rollups/RollupGenerator.java]: fixed

File: app/com/arpnetworking/rollups/RollupGenerator.java [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+                        return new TagNamesMessage.Builder()
+                                .setFailure(true)
+                                .setMetricName(metricName)
+                                .setThrowable(failure)
+                                .build();
+                    } else {
+                        return new TagNamesMessage.Builder()
+                                .setFailure(false)
+                                .setMetricName(metricName)
+                                .setTagNames(response.getQueries().get(0).getResults().get(0).getTags().keySet())
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/rollups/RollupGenerator.java]: fixed

File: app/com/arpnetworking/rollups/RollupGenerator.java [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+        final LastDataPointMessage.Builder builder = new LastDataPointMessage.Builder()
+                .setMetricName(metricName)
+                .setPeriod(period);
+        if (failure != null) {
+            builder.setFailure(true);
+            builder.setThrowable(failure);
+        } else {
+            builder.setFailure(false);
+            builder.setTags(tags);
+            builder.setLastDataPointTime(response.getQueries().get(0).getResults().get(0).getValues().get(0).getTime());
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/rollups/RollupGenerator.java]: fixed

File: app/com/arpnetworking/rollups/RollupGenerator.java (hunk 1 above)
Comment (ddimensia) [reply app/com/arpnetworking/rollups/RollupGenerator.java]: fixed

File: app/com/arpnetworking/rollups/LastDataPointMessage.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+    public RollupPeriod getPeriod() {
+        return _period;
+    }
+
+    private LastDataPointMessage(final Builder builder) {
+        super(builder);
+        _metricName = builder._metricName;
+        _tags = builder._tags;
+        _lastDataPointTime = builder._lastDataPointTime;
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/rollups/LastDataPointMessage.java]: Findbugs failed when using an optional as a field type.

File: app/com/arpnetworking/rollups/RollupGenerator.java [hunk 12]
```diff
@@ -0,0 +1,284 @@
...
+                                new Metric.Builder()
+                                        .setName(metricName)
+                                        .build()
+                        ))
+                        .build()).handle((response, failure) -> {
+                    if (failure != null) {
+                        return new TagNamesMessage.Builder()
+                                .setFailure(true)
+                                .setMetricName(metricName)
+                                .setThrowable(failure)
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/rollups/RollupGenerator.java]: Switched to using presence of throwable to determine isFailure.

File: app/com/arpnetworking/rollups/RollupGenerator.java [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+                getContext().dispatcher())
+                .to(getSelf());
+
+    }
+
+    private void handleTagNamesMessage(final TagNamesMessage message) {
+        if (message.isFailure()) {
+            // TODO(gmarkham): Metrics
+            LOGGER.warn(
+                    "Failed to get tag names for metric " + message.getMetricName(),
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/rollups/RollupGenerator.java]: fixed

File: app/com/arpnetworking/rollups/RollupGenerator.java [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+                                ), getContext().dispatcher())
+                        .to(getSelf());
+            }
+        }
+    }
+
+    private void handleLastDataPointMessage(final LastDataPointMessage message) {
+        if (message.isFailure()) {
+            // TODO(gmarkham): Metrics
+            LOGGER.warn(
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/rollups/RollupGenerator.java]: fixed

File: app/com/arpnetworking/rollups/RollupGenerator.java [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+
+    private void handleFinishRollupMessage(final FinishRollupMessage message) {
+        _periodsInFlight.remove(message.getPeriod());
+        if (_periodsInFlight.isEmpty()) {
+            getSelf().tell(FETCH_METRIC, ActorRef.noSender());
+        }
+    }
+
+    private LastDataPointMessage buildLastDataPointResponse(
+            final String metricName, final RollupPeriod period, final ImmutableSet<String> tags,
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/rollups/RollupGenerator.java]: fixed

File: app/com/arpnetworking/rollups/TagNamesMessage.java [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.rollups;
+
+import com.google.common.collect.ImmutableSet;
+import net.sf.oval.constraint.NotEmpty;
+import net.sf.oval.constraint.NotNull;
+
+/**
```
[the remark this answers is not in the document]
Comment (ddimensia) [reply app/com/arpnetworking/rollups/TagNamesMessage.java]: fixed

**Pull request merged**